### PR TITLE
Storage sandbox refactoring

### PR DESCRIPTION
**Pull request by LavrovArtem**

### CODE REVIEW

**Review by Farfurix — APPROVED** [review]

File: src/client/sandbox/storages/index.ts [hunk 1]
```diff
@@ -6,67 +6,79 @@ import { getTopSameDomainWindow } from '../../utils/dom';
...
+
+
+interface StoragesBackup {
+    localStorage: string;
+    sessionStorage: string;
+}
+
+const API_KEY_PREFIX = 'hammerhead|api-key-prefix|';
+const STORAGE_PROPS  = nativeMethods.arrayConcat.call(nativeMethods.objectKeys(Storage.prototype),
+    StorageWrapper.HH_INTERNAL_METHODS);
```
Comment (miherlosev) [inline src/client/sandbox/storages/index.ts:25]: `HH` is an unnecessary context.
Rename to `StorageWrapper.INTERNAL_METHODS`.

File: src/client/sandbox/storages/index.ts [hunk 2]
```diff
@@ -75,120 +87,181 @@ export default class StorageSandbox extends SandboxBase {
...
-        this._createStorageWrappers();
+            if (!storage.clearStorage())
+                return;
 
-        this.onLocalStorageChangeListener = this.localStorageWrapper.on(this.localStorageWrapper.STORAGE_CHANGED_EVENT,
-            e => this._simulateStorageEventIfNecessary(e, this.localStorageWrapper));
-        this.onSessionStorageListener     = this.sessionStorageWrapper.on(this.sessionStorageWrapper.STORAGE_CHANGED_EVENT,
-            e => this._simulateStorageEventIfNecessary(e, this.sessionStorageWrapper));
+            storage.raiseStorageChanged(null, null, null);
+            storage.internal.lastState = storage.getCurrentState();
```
Comment (miherlosev) [inline src/client/sandbox/storages/index.ts:127]: Restructure as the follows:
```js
storage.raiseStorageChanged(null, null, null);
storage.updateCurrentState();
```

File: src/client/sandbox/storages/wrapper.ts [hunk 3]
```diff
@@ -4,288 +4,227 @@ import { parseProxyUrl } from '../../utils/url';
...
 
-const API_KEY_PREFIX = 'hammerhead|api-key-prefix|';
-const KEY            = 0;
-const VALUE          = 1;
 
-function getWrapperMethods () {
-    const methods = [];
+const STORAGE_WRAPPER_KEY = 'hammerhead|get-storage-wrapper';
+const EMPTY_OLD_VALUE_ARG = isIE ? '' : null;
+const INTERNAL_WEAK_MAP   = new WeakMap<StorageWrapper, StorageInternalInfo>();
```
Comment (miherlosev) [inline src/client/sandbox/storages/wrapper.ts:12]: Bad name.
Compare with `internal_number`, `internal_boolean`.
What about the `storageWrappersInternalInfo` name?

**Review by miherlosev — CHANGES_REQUESTED** [review]
Excellent work. Just a few naming issues.

**Comment by LavrovArtem** [comment]
[testcafe-hammerhead-19.3.0.zip](https://github.com/DevExpress/testcafe-hammerhead/files/5910996/testcafe-hammerhead-19.3.0.zip)

**Review by miherlosev — APPROVED** [review]

**Comment by LavrovArtem** [comment]
https://github.com/DevExpress/testcafe/pull/5909

**Comment by LavrovArtem** [comment]
@testcafe-build-bot retest

**Pull request merged**